## Part 1: Preprocessing

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import numpy as np
from tensorflow.keras.models import Model
from tensorflow.keras import layers

#  Import and read the attrition data
attrition_df = pd.read_csv('https://static.bc-edx.com/ai/ail-v-1-0/m19/lms/datasets/attrition.csv')
attrition_df.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   41       Yes      Travel_Rarely                   Sales                 1   
1   49        No  Travel_Frequently  Research & Development                 8   
2   37       Yes      Travel_Rarely  Research & Development                 2   
3   33        No  Travel_Frequently  Research & Development                 3   
4   27        No      Travel_Rarely  Research & Development                 2   

   Education EducationField  EnvironmentSatisfaction  HourlyRate  \
0          2  Life Sciences                        2          94   
1          1  Life Sciences                        3          61   
2          2          Other                        4          92   
3          4  Life Sciences                        4          56   
4          1        Medical                        1          40   

   JobInvolvement  ...  PerformanceRating RelationshipSatisfaction  \
0               3  ...                  3                        1   
1               2  ...                  4                        4   
2               2  ...                  3                        2   
3               3  ...                  3                        3   
4               3  ...                  3                        4   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  \
0                 0                 8                      0               1   
1                 1                10                      3               3   
2                 0                 7                      3               3   
3                 0                 8                      3               3   
4                 1                 6                      3               3   

   YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
0               6                   4                        0   
1              10                   7                        1   
2               0                   0                        0   
3               8                   7                        3   
4               2                   2                        2   

   YearsWithCurrManager  
0                     5  
1                     7  
2                     0  
3                     0  
4                     2  

[5 rows x 27 columns]

In [2]:
# Determine the number of unique values in each column.
attrition_df.nunique()

Age                         43
Attrition                    2
BusinessTravel               3
Department                   3
DistanceFromHome            29
Education                    5
EducationField               6
EnvironmentSatisfaction      4
HourlyRate                  71
JobInvolvement               4
JobLevel                     5
JobRole                      9
JobSatisfaction              4
MaritalStatus                3
NumCompaniesWorked          10
OverTime                     2
PercentSalaryHike           15
PerformanceRating            2
RelationshipSatisfaction     4
StockOptionLevel             4
TotalWorkingYears           40
TrainingTimesLastYear        7
WorkLifeBalance              4
YearsAtCompany              37
YearsInCurrentRole          19
YearsSinceLastPromotion     16
YearsWithCurrManager        18
dtype: int64

In [33]:
# breakdown of Attrition column values
attrition_df['Attrition'].value_counts()

Attrition
No     1233
Yes     237
Name: count, dtype: int64

In [3]:
# Create y_df with the Attrition and Department columns 
y_df = attrition_df[['Attrition', 'Department']]
y_df.head()



Attrition              Department
0       Yes                   Sales
1        No  Research & Development
2       Yes  Research & Development
3        No  Research & Development
4        No  Research & Development

In [4]:
# Create a list of at least 10 column names to use as X data 
columns = ['Education','Age', 'OverTime', 'DistanceFromHome', 'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion', 'YearsWithCurrManager', 'JobSatisfaction', 'WorkLifeBalance']



# Create X_df using your selected columns

X_df = attrition_df[columns]

# Show the data types for X_df
X_df.dtypes


Education                   int64
Age                         int64
OverTime                   object
DistanceFromHome            int64
YearsAtCompany              int64
YearsInCurrentRole          int64
YearsSinceLastPromotion     int64
YearsWithCurrManager        int64
JobSatisfaction             int64
WorkLifeBalance             int64
dtype: object

In [5]:
# Split the data into training and testing sets
from sklearn.model_selection import train_test_split 
X_train, X_test, y_train, y_test = train_test_split(X_df, y_df, random_state=78)

X_train.info()



<class 'pandas.core.frame.DataFrame'>
Index: 1102 entries, 591 to 709
Data columns (total 10 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   Education                1102 non-null   int64 
 1   Age                      1102 non-null   int64 
 2   OverTime                 1102 non-null   object
 3   DistanceFromHome         1102 non-null   int64 
 4   YearsAtCompany           1102 non-null   int64 
 5   YearsInCurrentRole       1102 non-null   int64 
 6   YearsSinceLastPromotion  1102 non-null   int64 
 7   YearsWithCurrManager     1102 non-null   int64 
 8   JobSatisfaction          1102 non-null   int64 
 9   WorkLifeBalance          1102 non-null   int64 
dtypes: int64(9), object(1)
memory usage: 94.7+ KB


In [6]:
# Convert your X data to numeric data types however you see fit
# Add new code cells as necessary

# only OverTime is not numeric so we will convert it to numeric
X_train['OverTime'] = X_train['OverTime'].replace({'Yes': 1, 'No': 0})
X_test['OverTime'] = X_test['OverTime'].replace({'Yes': 1, 'No': 0})

X_train.info()


<class 'pandas.core.frame.DataFrame'>
Index: 1102 entries, 591 to 709
Data columns (total 10 columns):
 #   Column                   Non-Null Count  Dtype
---  ------                   --------------  -----
 0   Education                1102 non-null   int64
 1   Age                      1102 non-null   int64
 2   OverTime                 1102 non-null   int64
 3   DistanceFromHome         1102 non-null   int64
 4   YearsAtCompany           1102 non-null   int64
 5   YearsInCurrentRole       1102 non-null   int64
 6   YearsSinceLastPromotion  1102 non-null   int64
 7   YearsWithCurrManager     1102 non-null   int64
 8   JobSatisfaction          1102 non-null   int64
 9   WorkLifeBalance          1102 non-null   int64
dtypes: int64(10)
memory usage: 94.7 KB


/var/folders/b_/hfr4p_vs0nq_qyzqdx8n584m0000gn/T/ipykernel_94592/2647627121.py:5: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  X_train['OverTime'] = X_train['OverTime'].replace({'Yes': 1, 'No': 0})
/var/folders/b_/hfr4p_vs0nq_qyzqdx8n584m0000gn/T/ipykernel_94592/2647627121.py:6: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  X_test['OverTime'] = X_test['OverTime'].replace({'Yes': 1, 'No': 0})


In [7]:
# Create a StandardScaler 
scaler = StandardScaler()

# Fit the StandardScaler to the training data 
X_scaler = scaler.fit(X_train)


# Scale the training and testing data
X_train_scaled = X_scaler.transform(X_train)



In [8]:
# Create a OneHotEncoder for the Department column 
from sklearn.preprocessing import OneHotEncoder
enc = OneHotEncoder(sparse_output=False) 

# Fit the encoder to the training data
department_encoded = enc.fit(y_train[['Department']]) 

# Create two new variables by applying the encoder 
# to the training and testing data 
department_train = department_encoded.transform(y_train[['Department']])
department_test = department_encoded.transform(y_test[['Department']])
department_train[:5]




array([[0., 0., 1.],
       [0., 1., 0.],
       [0., 0., 1.],
       [0., 1., 0.],
       [0., 1., 0.]])

In [9]:
# Create a OneHotEncoder for the Attrition column 


# Fit the encoder to the training data 
attrition_encoded = enc.fit(y_train[['Attrition']])


# Create two new variables by applying the encoder 
# to the training and testing data 
attrition_train = attrition_encoded.transform(y_train[['Attrition']])
attrition_test = attrition_encoded.transform(y_test[['Attrition']])
attrition_train[:5]



array([[0., 1.],
       [1., 0.],
       [0., 1.],
       [1., 0.],
       [1., 0.]])

## Create, Compile, and Train the Model

In [16]:
# Find the number of columns in the X training data 

# Create the input layer 
input_layer = layers.Input(shape=(X_train.shape[1],), name = 'input')

# Create at least two shared layers
shared_layer1 = layers.Dense(64, activation='relu', name = 'shared1')(input_layer)
shared_layer2 = layers.Dense(32, activation='relu', name = 'shared2')(shared_layer1)



In [18]:
# Create a branch for Department
# with a hidden layer and an output layer 
department_branch = layers.Dense(16, activation='relu', name = 'dept')(shared_layer2) 


# Create the hidden layer 
department_hidden = layers.Dense(8, activation='relu', name = 'dept_hidden')(department_branch)


# Create the output layer 
department_output = layers.Dense(department_train.shape[1], activation='softmax', name='department')(department_hidden)




In [19]:
# Create a branch for Attrition
# with a hidden layer and an output layer 
attrition_branch = layers.Dense(16, activation='relu', name = 'attr')(shared_layer2)

# Create the hidden layer 
attrition_hidden = layers.Dense(8, activation='relu', name = 'attr_hidden')(attrition_branch)


# Create the output layer 
attrition_output = layers.Dense(units=attrition_train.shape[1], activation='softmax', name='attrition')(attrition_hidden)



In [22]:
# Create the model 
model = Model(inputs=input_layer, outputs=[department_output, attrition_output])


# Compile the model 
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy', 'accuracy'])


# Summarize the model 
model.summary()


Model: "functional_2"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input (InputLayer)  │ (None, 10)        │          0 │ -                 │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ shared1 (Dense)     │ (None, 64)        │        704 │ input[0][0]       │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ shared2 (Dense)     │ (None, 32)        │      2,080 │ shared1[0][0]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dept (Dense)        │ (None, 16)        │        528 │ shared2[0][0]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ attr (Dense)        │ (None, 16)        │        528 │ shared2[0][0]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dept_hidden (Dense) │ (None, 8)         │        136 │ dept[0][0]        │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ attr_hidden (Dense) │ (None, 8)         │        136 │ attr[0][0]        │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ department (Dense)  │ (None, 3)         │         27 │ dept_hidden[0][0] │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ attrition (Dense)   │ (None, 2)         │         18 │ attr_hidden[0][0] │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘

 Total params: 4,157 (16.24 KB)

 Trainable params: 4,157 (16.24 KB)

 Non-trainable params: 0 (0.00 B)

In [30]:
# Train the model 
model.fit(X_train_scaled,
            {'department': department_train, 'attrition': attrition_train},
             epochs=100,
             batch_size=32,
             validation_split=0.2, )



Epoch 1/100
28/28 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - attrition_accuracy: 0.9844 - attrition_loss: 0.0607 - department_accuracy: 0.9434 - department_loss: 0.1890 - loss: 0.2498 - val_attrition_accuracy: 0.7873 - val_attrition_loss: 0.8288 - val_department_accuracy: 0.5294 - val_department_loss: 1.5133 - val_loss: 2.3489
Epoch 2/100
28/28 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - attrition_accuracy: 0.9831 - attrition_loss: 0.0627 - department_accuracy: 0.9296 - department_loss: 0.1965 - loss: 0.2591 - val_attrition_accuracy: 0.7783 - val_attrition_loss: 0.8409 - val_department_accuracy: 0.5339 - val_department_loss: 1.5365 - val_loss: 2.3843
Epoch 3/100
28/28 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - attrition_accuracy: 0.9804 - attrition_loss: 0.0638 - department_accuracy: 0.9404 - department_loss: 0.1820 - loss: 0.2459 - val_attrition_accuracy: 0.7783 - val_attrition_loss: 0.9153 - val_department_accuracy: 0.5475 - val_department_loss: 1.6247 - val_loss: 2.5502
Epoch 4/100
28/28 ━━━━━━━━━━━━━━━━━━━

In [31]:
# Evaluate the model with the testing data
test_results = model.evaluate(X_test, {'department': department_test, 'attrition': attrition_test})
test_results

12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - attrition_accuracy: 0.3241 - attrition_loss: 87.1092 - department_accuracy: 0.4072 - department_loss: 42.3625 - loss: 129.5142 


[130.9219207763672,
 41.775997161865234,
 88.86941528320312,
 0.323369562625885,
 0.40489131212234497]

In [32]:
# Print the accuracy for both department and attrition 
print(f"Department predictions accuracy: {test_results[3]}")
print(f"Attrition predictions accuracy: {test_results[4]}")


Department predictions accuracy: 0.323369562625885
Attrition predictions accuracy: 0.40489131212234497


# Summary

In the provided space below, briefly answer the following questions.

1. Is accuracy the best metric to use on this data? Why or why not?

2. What activation functions did you choose for your output layers, and why?

3. Can you name a few ways that this model might be improved?

YOUR ANSWERS HERE

1. In the context of an employee attrition dataset, precision might be a better metric than accuracy. The Attrition column values show that the dataset is imbalanced (1233:No, 237:Yes), where the number of employees who stay is much larger than the number who leave. In such cases, accuracy can be misleading because a model that always predicts the majority class (e.g., predicting that all employees will stay) can still have high accuracy but not provide a true representation of the model's performance. Precision is the ratio of true positive predictions to the total number of positive predictions (true positives + false positives). It tells you how many of the employees predicted to leave actually left. This is crucial if the cost of a false positive (predicting an employee will leave when they won't) is high, such as unnecessary retention efforts or interventions. Therefore, precision provides a more meaningful evaluation metric in this context.

2. I used the `softmax` activation function for the output layers. The `softmax` activation function is appropriate in this context because it is typically used in the output layer of a neural network for multi-class classification problems. The `softmax` function converts the raw output scores (logits) from the network into probabilities that sum to 100%, making it easier to interpret the results. This is particularly useful when the model needs to classify inputs into multiple categories, such as different departments or attrition status. `softmax` ensures that the output values are in the range (0, 1) and sum up to 1, which is essential for probabilistic interpretation.

3. The model can be improved by increasing the size of the dataset, which can help the model learn better and generalize well to unseen data. Additionally, using precision as the evaluation metric instead of accuracy can provide a more accurate assessment of the model's performance, especially in the context of imbalanced datasets. Other potential improvements include experimenting with different neural network architectures, tuning hyperparameters, and incorporating additional relevant features into the model.